File: bonfire_lib/status.py
"""Reservation status queries, polling, and namespace description.

Replaces bonfire/openshift.py get_reservation(), get_all_reservations(),
wait_on_reservation(), get_console_url(), check_for_existing_reservation()
and bonfire/namespaces.py describe_namespace()
— using EphemeralK8sClient instead of ocviapy.
"""

import base64
import logging
import time

from bonfire_lib.k8s_client import EphemeralK8sClient
from bonfire_lib.utils import FatalError

log = logging.getLogger(__name__)


def get_reservation(
    client: EphemeralK8sClient,
    name: str | None = None,
    namespace: str | None = None,
    requester: str | None = None,
) -> dict | None:
    """Look up a reservation by name, namespace, or requester.

    Matches the lookup semantics of bonfire/openshift.py:get_reservation().
    """
    if name:
        return client.get_reservation(name)
    elif namespace:
        for res in client.list_reservations():
            if res.get("status", {}).get("namespace") == namespace:
                return res
    elif requester:
        reservations = client.list_reservations(label_selector=f"requester={requester}")
        if len(reservations) == 0:
            return None
        elif len(reservations) == 1:
            return reservations[0]
        else:
            log.info("Multiple reservations found for requester '%s'", requester)
            return None
    return None


def list_reservations(
    client: EphemeralK8sClient,
    requester: str | None = None,
) -> list[dict]:
    """List all reservations, optionally filtered by requester.

    Returns:
        List of reservation summary dicts.
    """
    if requester:
        reservations = client.list_reservations(label_selector=f"requester={requester}")
    else:
        reservations = client.list_reservations()

    result = []
    for res in reservations:
        status = res.get("status", {})
        spec = res.get("spec", {})
        result.append(
            {
                "name": res["metadata"]["name"],
                "namespace": status.get("namespace", ""),
                "state": status.get("state", ""),
                "expiration": status.get("expiration", ""),
                "requester": spec.get("requester", ""),
                "pool": spec.get("pool", "default"),
                "duration": spec.get("duration", ""),
            }
        )
    return result


def get_reservation_summary(res: dict) -> dict:
    """Extract a structured summary from a raw reservation dict.

    Avoids duplicating extraction logic across callers.
    """
    status = res.get("status", {})
    spec = res.get("spec", {})
    return {
        "name": res["metadata"]["name"],
        "namespace": status.get("namespace", ""),
        "state": status.get("state", ""),
        "expiration": status.get("expiration", ""),
        "requester": spec.get("requester", ""),
        "pool": spec.get("pool", "default"),
        "duration": spec.get("duration", ""),
    }


def wait_on_reservation(
    client: EphemeralK8sClient,
    name: str,
    timeout: int = 600,
) -> str:
    """Poll reservation until namespace is assigned.

    Returns:
        The assigned namespace name.

    Raises:
        TimeoutError if namespace not assigned within timeout.
    """
    log.info("waiting for reservation '%s' to get picked up by operator", name)
    start = time.time()
    while time.time() - start < timeout:
        res = client.get_reservation(name)
        if res:
            ns = res.get("status", {}).get("namespace")
            if ns:
                return ns
        time.sleep(2)
    raise TimeoutError(f"timed out after {timeout}s waiting for namespace on reservation '{name}'")


def check_for_existing_reservation(
    client: EphemeralK8sClient,
    requester: str,
) -> bool:
    """Check if requester already has an active reservation."""
    for res in client.list_reservations():
        if (
            res.get("spec", {}).get("requester") == requester
            and res.get("status", {}).get("state") == "active"
        ):
            ns = res["status"].get("namespace", "")
            if ns and client.get_namespace(ns):
                return True
    return False


def get_console_url(client: EphemeralK8sClient) -> str | None:
    """Get the OpenShift console URL from the cluster's console-public configmap."""
    try:
        cm = client.get_configmap("console-public", "openshift-config-managed")
        if cm:
            return cm.get("data", {}).get("consoleURL")
    except Exception as err:
        log.debug("unable to obtain console url: %s: %s", err.__class__.__name__, err)
    return None


def describe_namespace(client: EphemeralK8sClient, namespace: str) -> dict:
    """Get detailed information about an ephemeral namespace.

    Returns a dict with namespace details including:
    - namespace name, console URL
    - ClowdApp count and status
    - Frontend count
    - Gateway route, keycloak credentials
    """
    ns = client.get_namespace(namespace)
    if not ns:
        raise FatalError(f"namespace '{namespace}' not found")

    labels = ns.get("metadata", {}).get("labels", {})
    if labels.get("operator-ns") != "true":
        raise FatalError(f"namespace '{namespace}' was not reserved with namespace operator")

    try:
        clowdapps = client.list_crds("ClowdApp", namespace=namespace)
    except Exception as exc:
        log.warning("failed to list ClowdApps in namespace '%s': %s", namespace, exc)
        clowdapps = []

    try:
        frontends = client.list_crds("Frontend", namespace=namespace)
    except Exception as exc:
        log.warning("failed to list Frontends in namespace '%s': %s", namespace, exc)
        frontends = []

    fe_host = ""
    keycloak_url = ""
    try:
        fe_env = client.get_crd("FrontendEnvironment", f"env-{namespace}")
        if fe_env:
            fe_host = fe_env.get("spec", {}).get("hostname", "")
            keycloak_url = fe_env.get("spec", {}).get("sso", "")
    except Exception as exc:
        log.warning("failed to get FrontendEnvironment for namespace '%s': %s", namespace, exc)

    kc_creds = _get_keycloak_creds(client, namespace)

    console_url = get_console_url(client)
    ns_url = f"{console_url}/k8s/cluster/projects/{namespace}" if console_url else ""

    return {
        "namespace": namespace,
        "console_namespace_route": ns_url,
        "keycloak_admin_route": keycloak_url,
        "keycloak_admin_username": kc_creds.get("username", "N/A"),
        "keycloak_admin_password": kc_creds.get("password", "N/A"),
        "clowdapps_deployed": len(clowdapps),
        "frontends_deployed": len(frontends),
        "default_username": kc_creds.get("defaultUsername", "N/A"),
        "default_password": kc_creds.get("defaultPassword", "N/A"),
        "gateway_route": f"https://{fe_host}" if fe_host else "",
    }


def _get_keycloak_creds(client: EphemeralK8sClient, namespace: str) -> dict:
    """Get keycloak credentials from the namespace's keycloak secret."""
    secret = client.get_secret(f"env-{namespace}-keycloak", namespace)
    creds = {}
    if secret and "data" in secret:
        for key in ("username", "password", "defaultUsername", "defaultPassword"):
            raw = secret["data"].get(key, "")
            creds[key] = base64.b64decode(raw).decode("utf-8") if raw else "N/A"
    else:
        for key in ("username", "password", "defaultUsername", "defaultPassword"):
            creds[key] = "N/A"
    return creds

File: go/pkg/ephemeral/status.go
package ephemeral

import (
	"context"
	"encoding/base64"
	"fmt"
	"log/slog"
	"time"
)

// ReservationSummary is a flattened view of a NamespaceReservation CR.
type ReservationSummary struct {
	Name       string
	Namespace  string
	State      string
	Expiration string
	Requester  string
	Pool       string
	Duration   string
}

// GetReservation looks up a reservation by name, namespace, or requester.
// At most one parameter should be non-empty; returns nil, nil if not found.
// Mirrors bonfire_lib/status.py:get_reservation().
func GetReservation(ctx context.Context, client Client, name, namespace, requester string) (map[string]any, error) {
	switch {
	case name != "":
		return client.GetReservation(ctx, name)

	case namespace != "":
		all, err := client.ListReservations(ctx, "")
		if err != nil {
			return nil, err
		}
		for _, res := range all {
			if nestedString(res, "status", "namespace") == namespace {
				return res, nil
			}
		}
		return nil, nil

	case requester != "":
		all, err := client.ListReservations(ctx, "requester="+requester)
		if err != nil {
			return nil, err
		}
		if len(all) == 1 {
			return all[0], nil
		}
		if len(all) > 1 {
			slog.Info("multiple reservations found for requester", "requester", requester)
		}
		return nil, nil

	default:
		return nil, nil
	}
}

// GetReservationSummary extracts a structured summary from a raw reservation dict.
func GetReservationSummary(res map[string]any) ReservationSummary {
	return ReservationSummary{
		Name:       nestedString(res, "metadata", "name"),
		Namespace:  nestedString(res, "status", "namespace"),
		State:      nestedString(res, "status", "state"),
		Expiration: nestedString(res, "status", "expiration"),
		Requester:  nestedString(res, "spec", "requester"),
		Pool:       nestedString(res, "spec", "pool"),
		Duration:   nestedString(res, "spec", "duration"),
	}
}

// ListReservations lists all reservations, optionally filtered by requester.
// Mirrors bonfire_lib/status.py:list_reservations().
func ListReservations(ctx context.Context, client Client, requester string) ([]ReservationSummary, error) {
	var selector string
	if requester != "" {
		selector = "requester=" + requester
	}
	all, err := client.ListReservations(ctx, selector)
	if err != nil {
		return nil, err
	}
	out := make([]ReservationSummary, len(all))
	for i, res := range all {
		out[i] = GetReservationSummary(res)
	}
	return out, nil
}

// WaitOnReservation polls a reservation until status.namespace is populated.
// Returns the assigned namespace name.
// Returns an error if timeout is exceeded (caller should release the reservation).
// Mirrors bonfire_lib/status.py:wait_on_reservation().
func WaitOnReservation(ctx context.Context, client Client, name string, timeoutSecs int) (string, error) {
	slog.Info("waiting for reservation to be picked up by operator", "name", name)
	deadline := time.Now().Add(time.Duration(timeoutSecs) * time.Second)
	for time.Now().Before(deadline) {
		res, err := client.GetReservation(ctx, name)
		if err != nil {
			return "", err
		}
		if res != nil {
			if ns := nestedString(res, "status", "namespace"); ns != "" {
				return ns, nil
			}
		}
		select {
		case <-ctx.Done():
			return "", ctx.Err()
		case <-time.After(2 * time.Second):
		}
	}
	return "", fmt.Errorf("timed out after %ds waiting for namespace on reservation %q", timeoutSecs, name)
}

// CheckForExistingReservation returns true if the requester already has an active reservation
// with a live namespace.
// Mirrors bonfire_lib/status.py:check_for_existing_reservation().
func CheckForExistingReservation(ctx context.Context, client Client, requester string) (bool, error) {
	all, err := client.ListReservations(ctx, "")
	if err != nil {
		return false, err
	}
	for _, res := range all {
		if nestedString(res, "spec", "requester") == requester &&
			nestedString(res, "status", "state") == "active" {
			ns := nestedString(res, "status", "namespace")
			if ns == "" {
				continue
			}
			nsObj, err := client.GetNamespace(ctx, ns)
			if err != nil {
				return false, err
			}
			if nsObj != nil {
				return true, nil
			}
		}
	}
	return false, nil
}

// GetConsoleURL returns the OpenShift console URL from the cluster's console-public ConfigMap.
// Returns "" if not available.
// Mirrors bonfire_lib/status.py:get_console_url().
func GetConsoleURL(ctx context.Context, client Client) string {
	cm, err := client.GetConfigMap(ctx, "console-public", "openshift-config-managed")
	if err != nil {
		slog.Debug("unable to obtain console url", "err", err)
		return ""
	}
	if cm == nil {
		return ""
	}
	return cm.Data["consoleURL"]
}

// NamespaceDescription holds detailed information about an ephemeral namespace.
type NamespaceDescription struct {
	Namespace              string
	ConsoleNamespaceRoute  string
	KeycloakAdminRoute     string
	KeycloakAdminUsername  string
	KeycloakAdminPassword  string
	ClowdAppsDeployed      int
	FrontendsDeployed      int
	DefaultUsername        string
	DefaultPassword        string
	GatewayRoute           string
	HasCluster             bool
}

// DescribeNamespace returns detailed information about an ephemeral namespace.
// Mirrors bonfire_lib/status.py:describe_namespace().
func DescribeNamespace(ctx context.Context, client Client, namespace string) (*NamespaceDescription, error) {
	ns, err := client.GetNamespace(ctx, namespace)
	if err != nil {
		return nil, fmt.Errorf("get namespace: %w", err)
	}
	if ns == nil {
		return nil, NewFatalError("namespace %q not found", namespace)
	}
	if ns.Labels["operator-ns"] != "true" {
		return nil, NewFatalError("namespace %q was not reserved with namespace operator", namespace)
	}

	clowdApps, err := client.ListCRDs(ctx, gvrClowdApp, namespace)
	if err != nil {
		slog.Warn("failed to list ClowdApps", "namespace", namespace, "err", err)
		clowdApps = nil
	}

	frontends, err := client.ListCRDs(ctx, gvrFrontend, namespace)
	if err != nil {
		slog.Warn("failed to list Frontends", "namespace", namespace, "err", err)
		frontends = nil
	}

	var feHost, keycloakURL string
	feEnv, err := client.GetCRD(ctx, gvrFrontendEnvironment, "env-"+namespace, "")
	if err != nil {
		slog.Warn("failed to get FrontendEnvironment", "namespace", namespace, "err", err)
	} else if feEnv != nil {
		feHost = nestedString(feEnv, "spec", "hostname")
		keycloakURL = nestedString(feEnv, "spec", "sso")
	}

	kc := getKeycloakCreds(ctx, client, namespace)
	consoleURL := GetConsoleURL(ctx, client)
	nsURL := ""
	if consoleURL != "" {
		nsURL = consoleURL + "/k8s/cluster/projects/" + namespace
	}

	hasCluster := hasClusterKubeconfig(ctx, client, namespace)

	gwRoute := ""
	if feHost != "" {
		gwRoute = "https://" + feHost
	}

	return &NamespaceDescription{
		Namespace:             namespace,
		ConsoleNamespaceRoute: nsURL,
		KeycloakAdminRoute:    keycloakURL,
		KeycloakAdminUsername: kc["username"],
		KeycloakAdminPassword: kc["password"],
		ClowdAppsDeployed:     len(clowdApps),
		FrontendsDeployed:     len(frontends),
		DefaultUsername:       kc["defaultUsername"],
		DefaultPassword:       kc["defaultPassword"],
		GatewayRoute:          gwRoute,
		HasCluster:            hasCluster,
	}, nil
}

func hasClusterKubeconfig(ctx context.Context, client Client, namespace string) bool {
	secret, err := client.GetSecret(ctx, namespace+"-cluster-kubeconfig", namespace)
	return err == nil && secret != nil
}

func getKeycloakCreds(ctx context.Context, client Client, namespace string) map[string]string {
	result := map[string]string{
		"username":        "N/A",
		"password":        "N/A",
		"defaultUsername": "N/A",
		"defaultPassword": "N/A",
	}
	secret, err := client.GetSecret(ctx, "env-"+namespace+"-keycloak", namespace)
	if err != nil || secret == nil {
		return result
	}
	for _, key := range []string{"username", "password", "defaultUsername", "defaultPassword"} {
		raw, ok := secret.Data[key]
		if !ok || len(raw) == 0 {
			continue
		}
		// Secret data is already []byte (base64-decoded by the API server).
		// But in the Python implementation it is double-decoded, so match that.
		decoded, err := base64.StdEncoding.DecodeString(string(raw))
		if err != nil {
			// Already raw bytes from API server
			result[key] = string(raw)
		} else {
			result[key] = string(decoded)
		}
	}
	return result
}
